File: android/src/main/java/com/margelo/nitro/lunardatepicker/LunarDatePicker.kt
package com.margelo.nitro.lunardatepicker

// Module imports
import android.util.Log
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.lunardatepicker.constants.DataConstants
import com.margelo.nitro.lunardatepicker.coordinators.LunarDatePickerCoordinator
import com.margelo.nitro.lunardatepicker.exceptions.LunarDatePickerException
import com.margelo.nitro.lunardatepicker.services.ConfigurationCacheService
import com.margelo.nitro.lunardatepicker.utils.DateConverter
import com.margelo.nitro.lunardatepicker.utils.MemoryOptimizer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import com.margelo.nitro.NitroModules

@DoNotStrip
class LunarDatePicker : HybridLunarDatePickerSpec() {

  companion object {
    private const val TAG = DataConstants.LogTags.MAIN
  }

  // MARK: - Properties
  private val coordinator: LunarDatePickerCoordinator by lazy {
    LunarDatePickerCoordinator()
  }
  private val dateConverter: DateConverter by lazy {
    DateConverter()
  }
  private val configurationCacheService: ConfigurationCacheService by lazy {
    ConfigurationCacheService()
  }

  // MARK: - HybridLunarDatePickerSpec Implementation

  /**
   * Configures the global settings for the date picker
   * @param config Configuration parameters including themes, languages, etc.
   */
  override fun configure(config: LDP_ConfigParams) {
    try {
      

      // Apply configuration to coordinator
      coordinator.configure(config)

      // Initialize performance optimizations
      initializePerformanceOptimizations()

      Log.d(TAG, "Configuration applied successfully")
    } catch (e: Exception) {
      Log.e(TAG, "Failed to configure: ${e.message}", e)
      throw LunarDatePickerException.InvalidConfiguration(e.message ?: "Invalid configuration")
    }
  }

  /**
   * Presents the date picker with the specified parameters
   * @param params Presentation parameters including mode, theme, language, etc.
   */
  override fun present(params: LDP_PresentParams) {
    try {
      // Validate presentation parameters
      validatePresentationParams(params)

      // Ensure we're presenting on main thread
      CoroutineScope(Dispatchers.Main).launch {
        coordinator.present(params)
      }

      Log.d(TAG, "Date picker presented successfully")
    } catch (e: Exception) {
      Log.e(TAG, "Failed to present: ${e.message}", e)
      throw LunarDatePickerException.PresentationFailed(e.message ?: "Presentation failed")
    }
  }

  /**
   * Cập nhật giá cho một tháng cụ thể
   * Có thể gọi khi calendar đang mở để cập nhật UI ngay lập tức
   */
  override fun updatePrices(params: LDP_PriceUpdateParams) {
    try {
      coordinator.updatePrices(params)
      Log.d(TAG, "Prices updated successfully")
    } catch (e: Exception) {
      Log.w(TAG, "Failed to update prices: ${e.message}", e)
    }
  }

  // MARK: - Private Methods

  /**
   * Initializes performance optimizations including object pooling and caching
   */
  private fun initializePerformanceOptimizations() {
    try {
      // Preload object pools for better memory management
      MemoryOptimizer.preloadObjectPools()

      val configuredTimeZone = coordinator.getConfiguredTimeZone()

      // Preload lunar date cache for faster calculations
      MemoryOptimizer.preloadLunarDateCache(dateConverter, configuredTimeZone ?: java.time.ZoneId.systemDefault())

      Log.d(TAG, "Performance optimizations initialized")
    } catch (e: Exception) {
      Log.w(TAG, "Failed to initialize performance optimizations", e)
      // Continue without optimizations - don't fail configuration
    }
  }

  /**
   * Validates the presentation parameters
   * @param params Presentation parameters to validate
   * @throws LunarDatePickerException.InvalidConfiguration if validation fails
   */
  private fun validatePresentationParams(params: LDP_PresentParams) {
    // Get timezone from coordinator's configuration for validation
    val timeZone = coordinator.getConfiguredTimeZone()
    
    // Validate date ranges if provided
    val minimumDateString = params.minimumDate
    val maximumDateString = params.maximumDate

    if (minimumDateString != null && maximumDateString != null) {
      val minimumDate = dateConverter.dateFromString(minimumDateString, timeZone)
      val maximumDate = dateConverter.dateFromString(maximumDateString, timeZone)

      if (minimumDate == null) {
        throw LunarDatePickerException.InvalidConfiguration("Invalid minimum date format. Expected DD/MM/YYYY")
      }
      if (maximumDate == null) {
        throw LunarDatePickerException.InvalidConfiguration("Invalid maximum date format. Expected DD/MM/YYYY")
      }

      if (minimumDate >= maximumDate) {
        throw LunarDatePickerException.InvalidConfiguration("Minimum date must be less than maximum date")
      }
    }

    // Validate initial value if provided
    params.initialValue?.let { initialValue ->
      val fromDate = dateConverter.dateFromString(initialValue.from, timeZone)
      if (fromDate == null) {
        throw LunarDatePickerException.InvalidConfiguration("Invalid initial value 'from' date format. Expected DD/MM/YYYY")
      }

      initialValue.to?.let { toDateString ->
        val toDate = dateConverter.dateFromString(toDateString, timeZone)
        if (toDate == null) {
          throw LunarDatePickerException.InvalidConfiguration("Invalid initial value 'to' date format. Expected DD/MM/YYYY")
        }
        if (fromDate > toDate) {
          throw LunarDatePickerException.InvalidConfiguration("Initial value 'from' date must be less than or equal to 'to' date")
        }
      }
    }
  }
}

File: ios/LunarDatePicker.swift
//
//  LunarDatePicker.swift
//  LunarDatePicker
//
//  Main entry point for React Native Nitro Module
//

import Foundation
import UIKit

/// Main class implementing the HybridLunarDatePickerSpec protocol
/// This class serves as the bridge between React Native and native iOS implementation
final class LunarDatePicker: HybridLunarDatePickerSpec {

  // MARK: - Properties

  /// Coordinator responsible for managing the presentation flow
  private let coordinator: LunarDatePickerCoordinator
  private let dateConverter: DateConverting
  private var currentConfig: LDP_ConfigParams?

  // MARK: - Initialization

  /// Initializes the LunarDatePicker with default dependencies
  override init() {
    self.dateConverter = DateConverter()
    self.coordinator = LunarDatePickerCoordinator()
  }

  /// Initializes the LunarDatePicker with custom dependencies (for testing)
  /// - Parameter coordinator: Custom coordinator instance
  init(coordinator: LunarDatePickerCoordinator) {
    self.dateConverter = DateConverter()
    self.coordinator = coordinator
  }

  // MARK: - HybridLunarDatePickerSpec Implementation

  /// Configures the global settings for the date picker
  /// - Parameter config: Configuration parameters including themes, languages, etc.
  /// - Throws: An error if configuration is invalid
  public func configure(config: LDP_ConfigParams) throws {
    // Validate configuration before applying
    try validateConfiguration(config)

    // Apply configuration to coordinator
    coordinator.configure(with: config)
    currentConfig = config
  }

  /// Presents the date picker with the specified parameters
  /// - Parameter params: Presentation parameters including theme, language, etc.
  /// - Throws: An error if presentation fails
  public func present(params: LDP_PresentParams) throws {
    // Ensure we're on the main thread for UI operations
    guard Thread.isMainThread else {
      DispatchQueue.main.async { [weak self] in
        try? self?.present(params: params)
      }
      return
    }

    // Validate presentation parameters
    try validatePresentationParams(params)

    // Present using coordinator
    try coordinator.present(with: params)
  }

  /// Cập nhật giá cho một tháng cụ thể
  public func updatePrices(params: LDP_PriceUpdateParams) throws {
    coordinator.updatePrices(with: params)
  }

  /// Cập nhật maximumDate khi calendar đang mở
  public func updateMaximumDate(maximumDate: String) throws {
    coordinator.updateMaximumDate(maximumDate)
  }


  // MARK: - Private Methods

  /// Validates the configuration parameters
  /// - Parameter config: Configuration to validate
  /// - Throws: LunarDatePickerError.invalidConfiguration if validation fails
  private func validateConfiguration(_ config: LDP_ConfigParams) throws {
    // Validate themes
    for (_, theme) in config.themes {
      guard isValidHexColor(theme.backgroundColor),
        isValidHexColor(theme.titleColor),
        isValidHexColor(theme.dateLabelColor),
        isValidHexColor(theme.todayLabelColor),
        isValidHexColor(theme.lunarDateLabelColor),
        isValidHexColor(theme.selectedTextColor),
        isValidHexColor(theme.weekendLabelColor),
        isValidHexColor(theme.specialDayLabelColor),
        isValidHexColor(theme.priceLabelColor),
        isValidHexColor(theme.cheapestPriceLabelColor),
        isValidHexColor(theme.monthLabelColor),
        isValidHexColor(theme.secondColor),
        isValidHexColor(theme.weekViewBackgroundColor),
        isValidHexColor(theme.selectedBackgroundColor),
        isValidHexColor(theme.rangeBackgroundColor),
        isValidHexColor(theme.submitButtonColor),
        isValidHexColor(theme.noticeLabelColor),
        isValidHexColor(theme.noticeBackgroundColor)
      else {
        throw LunarDatePickerError.invalidConfiguration
      }
    }

    // Validate languages
    for (_, language) in config.languages {
      guard language.weekdayNames.count == 7
      else {
        throw LunarDatePickerError.invalidConfiguration
      }
    }
  }

  /// Validates the presentation parameters
  /// - Parameter params: Presentation parameters to validate
  /// - Throws: LunarDatePickerError.invalidConfiguration if validation fails
  private func validatePresentationParams(_ params: LDP_PresentParams) throws {
    let timeZone = currentConfig.flatMap { config in
      TimeZone(secondsFromGMT: Int(config.timeZoneOffset * 3600))
    }

    func parse(_ value: String) throws -> Date {
      guard let date = dateConverter.dateFromString(value, timeZone: timeZone) else {
        throw LunarDatePickerError.invalidConfiguration
      }
      return date
    }

    if let minimumDateString = params.minimumDate,
       let maximumDateString = params.maximumDate {
      let minDate = try parse(minimumDateString)
      let maxDate = try parse(maximumDateString)
      guard minDate <= maxDate else { throw LunarDatePickerError.invalidConfiguration }
    }

    if let initialValue = params.initialValue {
      let fromDate = try parse(initialValue.from)
      if let toString = initialValue.to {
        let toDate = try parse(toString)
        guard fromDate <= toDate else { throw LunarDatePickerError.invalidConfiguration }
      }
    }
  }

  /// Checks if a string is a valid hex color
  /// - Parameter hex: Hex color string to validate
  /// - Returns: true if valid hex color, false otherwise
  private func isValidHexColor(_ hex: String) -> Bool {
    let hexRegex = "^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{8})$"
    let hexPredicate = NSPredicate(format: "SELF MATCHES %@", hexRegex)
    return hexPredicate.evaluate(with: hex)
  }
}
